src/mlpack/methods/ann/layer_names.hpp

```diff
@@ -0,0 +1,704 @@
+/**
+ * @file layer_names.hpp
+ * @author Sreenik Seal
+ *
+ * Implementation of a class that converts a given ann layer to string format.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/layer/layer_types.hpp>
+#include <boost/variant/static_visitor.hpp>
+#include <string>
+
+using namespace mlpack::ann;
+
+class TestVisitor : public boost::static_visitor<std::string>
+{
+ public:
+  TestVisitor() { /*Nothing to do here*/ }
+
+  template<typename LayerType>
+  std::string operator()(LayerType* layer) const
+  {
+    return LayerString(layer);
+  }
+
+ private:
+  template<typename T>
+  typename std::enable_if<
+      std::is_same<T, Linear<> >::value &&
+      !std::is_same<T, Convolution<> >::value &&
+      !std::is_same<T, ReLULayer<> >::value &&
+      !std::is_same<T, MaxPooling<> >::value &&
+      !std::is_same<T, AtrousConvolution<> >::value &&
+      !std::is_same<T, AlphaDropout<> >::value &&
+      !std::is_same<T, BatchNorm<> >::value &&
+      !std::is_same<T, Constant<> >::value &&
+      !std::is_same<T, DropConnect<> >::value &&
+      !std::is_same<T, Dropout<> >::value &&
+      !std::is_same<T, LayerNorm<> >::value &&
+      !std::is_same<T, LinearNoBias<> >::value &&
+      !std::is_same<T, MeanPooling<> >::value &&
+      !std::is_same<T, MultiplyConstant<> >::value &&
+      !std::is_same<T, TransposedConvolution<> >::value &&
+      !std::is_same<T, IdentityLayer<> >::value &&
+      !std::is_same<T, TanHLayer<> >::value &&
+      !std::is_same<T, ELU<> >::value &&
+      !std::is_same<T, HardTanH<> >::value &&
+      !std::is_same<T, LeakyReLU<> >::value &&
+      !std::is_same<T, PReLU<> >::value &&
+      !std::is_same<T, SigmoidLayer<> >::value &&
+      !std::is_same<T, LogSoftMax<> >::value,
+      std::string>::type
+  LayerString(T* layer) const { return "linear"; }
```

Comment:
On a second look we should be able to simplify this expression, by using:

```c++
LayerString(Linear<> layer) const { return "linear"; }
LayerString(Convolution <> layer) const { return "convolution"; }
```

Reply:
I don't think that's possible because not all layers are implemented yet so wouldn't the compiler be unable to determine if the LayerTypes<> instances obtained from the FFN model actually have functions with the respective types overloaded?

I am also unsure about error-free passing of LayerTypes<> objects to these methods directly. Let me know what you think.

Reply:
I was thinking about:

```c++
class LayerNameVisitor : public boost::static_visitor<std::string>
{
 public:
  LayerNameVisitor()
  {
  }

  std::string LayerString(LinearNoBias<>* /*layer*/) const
  {
    return "linearnobias";
  }

  template<typename T>
  std::string LayerString(T* /*layer*/) const
  {
    return "unknown";
  }

  template<typename LayerType>
  std::string operator()(LayerType* layer) const
  {
    return LayerString(layer);
  }
};
```

Every layer that isn't implemented is catched by: 

```c++
template<typename T>
std::string LayerString(T* /*layer*/) const
{
  return "unknown";
}
```

Reply:
Oh yes got it, this will work. I'll make the changes and push it. Thanks.